Make sidebar-notice JITM template dismissable

dsas opened the pull request:
Sidebar-notice JITMs generated on the server were not dismissable - they weren't rendered as dismissed and recording their dismissal was not done correctly.

They weren't rendered as dismissed as the JITM API response includes `isDismissible` to indicate
that a JITM ought to be rendered as dismissable but the SidebarBannerTemplate was using `dismissPreferenceName` instead, which isn't part of the JITM REST API.

The rendering issue was addressed by using isDismissable to generate a dismissPreferenceName and passing that to the generated Banner

Recording their dismissal worked on a calypso level - it is implemented in `DismissableCard` which persists to the `calypso_preferences` user attrib, this meant that the user wouldn't see the same message again . The JITM generating server code doesn't know about `calypso_preferences` though, it stores dismissals differently. This meant that the server JITM generator code was always returning the same top-ranked JITM and the `DismissableCard` was refusing to render itself due to being dismissed. 

The dismissal recording issue was addressed by passing onDismiss from UpsellNudge to Banner, rather than onDismissClick and meant that the dismissal was recorded in two places - once in `calypso_preferences` via [`DismissableCard`](https://github.com/Automattic/wp-calypso/blob/trunk/client/blocks/dismissible-card/index.jsx), which records dismissals based upon the JITM ID and once in the [Jetpack option `hide_jitm`](https://github.com/Automattic/jetpack/blob/6510767762d7415cfc0bd1fbecd5fed1b995d9c7/projects/packages/jitm/src/class-jitm.php#L274) via the [jitm calypso block](https://github.com/Automattic/wp-calypso/blob/75e663a3d0133b311a83267ccf735171a1935dec/client/blocks/jitm/index.jsx#L73) which records based upon the JITM feature_class. The `calypso_preferences` dismissal recording was then removed by passing the `dismissTemporary` property around.



https://user-images.githubusercontent.com/93301/141107266-6c2771dc-f6ca-4064-9803-7252b385c3de.mp4


#### Changes proposed in this Pull Request

 * Change SidebarBannerTemplate to:
   * Expect `onDismiss` rather than `onDismissClick`, [as this is what is passed to it](https://github.com/Automattic/wp-calypso/blob/75e663a3d0133b311a83267ccf735171a1935dec/client/blocks/jitm/index.jsx#L19)
   * Expect tracksDismissName to be passed in as part of the tracks variable rather than directly, and give it a default value
   * Generate `dismissPreferenceName` if the message `isDismissable` rather than expecting it to be passed in     
 * Change UpsellNudge to pass onDismiss to the Banner it creates, rather than on onDismissClick 

#### Testing instructions

1. Make your site eligible for two server-generated JITMs with the same feature_class and one in a different feature class
   * The easiest way to do this is to open `jitm-engine.php` in your wpcom sandbox and change the eligibility rules & feature_classes for messages that are defined using `with_template( 'sidebar-banner' );`
 2. Load up Calypso using one of the links below
 3. Note that the message appears without a dismiss button like in 'before' below
 4. Change the messages in `jitm-engine.php` in your wpcom sandbox so that they are dismissable (remove a line saying `is_dismissable( false )`)
 5. Add a tracks event to the messages too, something like `with_tracks( 'dismiss', 'calypso_yourname_test', [ 'cta_name' => 'foo-bar-yourname' ] );`
 5. Reload calypso, the same JITM should now appear with a X icon to dismiss it like in 'after' below
 6. Press X to dismiss it
 7. The message should disappear immediately
 8. Reload the page, the message should still not appear but one of the messages with a different feature class should
 9. Press X to dismiss it, it should disappear immedately
 10. Reload the page, none of your altered messages should appear.
 11. Check tracks, you should be able to see a 'calypso_yourname_test' event after 5-10 minutes


Before | After
-------|------
 ![image](https://user-images.githubusercontent.com/93301/140935132-457440fc-69c4-4a9a-bb3f-be2831cddad6.png) | ![image](https://user-images.githubusercontent.com/93301/140935161-5c510e5e-f41e-46a0-8bde-6000341fa7dc.png)


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #52368

dsas committed:
Make sidebar-notice JITM template dismissable

The sidebar-notice JITMs were not actually able to be marked as
dismissable via the REST API. The API uses isDismissible to indicate
that a JITM ought to be dismissed but this calypso component expected to
use dismissPreferenceName for that, which the API does not return.

dsas committed:
Use jetpack to persist dismissed JITM messages

The JITM functionality expects dismissed JITM messages to be recorded in
a particular way, which is done by POSTing to the jetpack/v4/jitm API
endpoint. This wasn't working for upsell nudges, however it was
recording the dismissal in calypso preferences.

Th impact of this was that the JITM API didn't know it was supposed to
be hiding a particular JITM and so kept sending it (when potentially it
could be sending another one). Calypso knew that that particular JITM
was not supposed to be displayed and so wouldn't display it.

This commit corrects the problem by having <UpsellNudge/> pass the
onDismiss property to <Banner/> rather than passing the onDismissClick
property, which Banner doesn't recognise.

BogdanUngureanu commented on client/blocks/jitm/templates/sidebar-banner.jsx:
```diff
@@ -3,20 +3,16 @@ import { preventWidows } from 'calypso/lib/formatting';
 	const displayName = tracks?.display?.name ?? `jitm_nudge_impression_${ id }`;
 	const displayProps = tracks?.display?.props;
+
+	const dismissName = tracks?.dismiss?.name ?? `jitm_nudge_dismiss_${ id }`;
```
Shouldn't we use `calypso_upgrade_nudge` and `calypso_upgrade_nudge_impression` so that it's consistent with jitm-engine.php 🤔 ? The dismissProps should contain the `cta_name` property to differentiate between the nudges.

dsas replied:
If something is specified in jitm-engine.php then that will be used in preference to the fallback dismissName set here. 

In jitm-engine it uses names like `calypso_upgrade_nudge_impression`, `calypso_upgrade_nudge_click`

In calypso, the fallbacks in this file are all of the form `jitm_nudge_dismiss_${ id }`

I'm not very knowledgeable about tracks, so I've based this on what the existing code does. I'm not sure why there is a disparity, but I just followed the existing pattern assuming it was done for a good reason. I will go back to my changes in D69690-code and add explicit dismiss tracks there.

dsas replied:
After further discussion and discovery, these default names just aren't going to work. I'm just going to leave this as it is. I've filed #57955 to document the issue and people will get a browser console message telling them it's not going to send the event to tracks.

I've made a comment on my wpcom review which uses the dismiss functionality to remind me to specify a dismiss name there.

dsas committed:
Selectively forceHref in sidebar-banner` JITM template

forceHref is a property passed down to <Banner> which forces it to
render the whole displayed card as a link, rather than just the button.
This was hardcoded to true in sidebar-banner, however this was not
compatible with dismissable sidebars - the whole card would correctly
not get marked as a link but the button also wouldn't get marked as a
link.

The impact of this was that you could read a sidebar-banner message, you
could dismiss it, but you couldn't do anything else with it.

This has been addressed by setting forceHref to false for dismissable
sidebar-banner messages. As a further aid to future developers
<UpsellNudge> will also emit a console message if it is given a truthy
forceHref and is to dismissable.

Relates to #57832

dsas committed:
Don't persist calypso dismissals for server-originated JITMs

Don't persist dismissals of server-originated sidebar
just-in-time-messages inside calypso user attributes on the server. They
have built-in dismissal tracking via their own endpoint, with its own
logic about retrying dismissals after a given amount of time and that
should be preferred.

mmtr commented on client/blocks/jitm/templates/sidebar-banner.jsx:
```diff
@@ -3,39 +3,44 @@ import { preventWidows } from 'calypso/lib/formatting';
-	tracksDismissName,
-	tracksDismissProperties,
-} ) {
+export default function SidebarBannerTemplate( { CTA, message, id, onDismiss, tracks, ...props } ) {
```
[In the past](https://github.com/Automattic/wp-calypso/pull/38881#discussion_r370402507), we've recommended to maintain explicit props rather than allowing arbitrary props. 

Seems we only check `props.isDismissible` below, so it may be better to note it here:
```suggestion
export default function SidebarBannerTemplate( { CTA, message, id, onDismiss, tracks, isDismissible = false } ) {
```

mmtr commented on client/blocks/upsell-nudge/index.jsx:
```diff
@@ -104,6 +108,12 @@ export const UpsellNudge = ( {
 		{ 'is-wpcom-plan': plan && planMatches( plan, { group: GROUP_WPCOM } ) }
 	);
 
+	if ( dismissPreferenceName && forceHref && href ) {
+		debug(
+			"It's not possible to force the whole banner to be a link when it needs to link and be dismissable"
+		);
+	}
+
```
Wouldn't make more sense to place this in `SidebarBannerTemplate` which is the place where we ignore the `forceHref` prop if the nudge is dimissible?

dsas replied:
Maybe this could be improved but I'm not sure this is it. `SidebarBannerTemplate` doesn't receive forceHref as a prop, it just gives a sane forceHref prop to `UpsellNudge`, which logs the above message in case someone else ever gives it an incompatible set of props.

It's actually the `Banner` components getContent method which does the ignoring, if called to provide the content for a `DismissibleCard`, so perhaps that ought to have a debug message too.

mmtr reviewed (commented):
Not sure if I'm testing something wrong, but I modified the `jitm-engine.php` file to make sure that my site is eligible for the `free_to_paid_plan`, `domain_to_paid_plan`, and `blogger_to_personal_plan` JITMs.

I made the first one dismissible, and then dismissed it on Calypso. However, after reloading Calypso, none of the other 2 JITMs show up.

dsas commented:
> I modified the jitm-engine.php file to make sure that my site is eligible for the free_to_paid_plan, domain_to_paid_plan, and blogger_to_personal_plan JITMs.
>
> I made the first one dismissible, and then dismissed it on Calypso. However, after reloading Calypso, none of the other 2 JITMs show up.

Did you change the `feature_class` (second parameter to Message constructor) on one of them? If not that is (sort of) as expected.

I **think** that the idea is that if you dismiss a nudge, you also dismiss very similar nudges. Similar nudges are determined by the feature_class of a message and all of those messages have the same feature_class.

I guess that as these messages are not dismissible, no-one thought too hard about the `feature_class` as it wasn't used for anything.

mmtr commented:
> I think that the idea is that if you dismiss a nudge, you also dismiss very similar nudges. Similar nudges are determined by the feature_class of a message and all of those messages have the same feature_class.
>
> I guess that as these messages are not dismissible, no-one thought too hard about the feature_class as it wasn't used for anything.

TIL! That makes sense, thanks for extra context.

mmtr commented:
Also noted that we're tracking the dismiss event twice:

<img width="726" alt="Screen Shot 2021-11-22 at 17 00 48" src="https://user-images.githubusercontent.com/1233880/142894348-d657db92-7893-4e48-82ef-56f0aad2c5ff.png">

I think the first one is triggered here:

https://github.com/Automattic/wp-calypso/blob/02b621f7bf67a2cdb4353bd1e05e5a9cca0a1bb9/client/blocks/jitm/index.jsx#L74-L75

And the other one here:
https://github.com/Automattic/wp-calypso/blob/02b621f7bf67a2cdb4353bd1e05e5a9cca0a1bb9/client/components/banner/index.jsx#L132-L136

According to how other JITM templates are implemented, it seems that we should rely on the handlers defined in `getEventHandlers`, and stop setting the `tracksClickName`, `tracksClickProperties`, `tracksImpressionName`, `tracksImpressionProperties`, `tracksDismissName`, and `tracksDismissProperties` props.

dsas committed:
Use explicit props rather than implicit

dsas committed:
Correct tracks events for JITM SidebarBanner

When dismissing a SidebarBanner the events was tracking twice - once for
the JITM specific tracking and once for the tracking that is part of the
Banner component. The clicks and dismiss events were using the Banner
component to trigger them, rather than the JITM code, though it was
implemented compatibly.

This change simplifies SidebarBanner and brings it further in-line with
other JITM templates by relying on the jitm specific logic more.

h/t @mmtr

mmtr reviewed (approved):
Works great! :shipit:

dsas commented:
Thanks for your help and review @mmtr, I'll ship this tomorrow morning.
